### PR TITLE
Improve BehaviourTest in java template [extern/old]

### PR DESCRIPTION
**Pull request by alllex**
### Motivation and Context
BehaviorTest class in templates for Java exercises had bugs and did not provide required functionality for one of the exercises.

### Description
First of all, there was a bug in the method `getParameterTypes` where it would not construct the correct array of classes if any of the parameters were equal in the Java-object sense.
Secondly, there was no way of creating a `newInstance` if the required classes had constructor with primitively-typed arguments. Method `getConstructor` and another overloading of `newInstance` were added to tackle that problem.
Next, there was no way of using `invokeMethod`, but rethrowing an exception which could occur during the underlying method execution. It is sometimes required to check whether implementation of a method by students throws an expected exception. A new method `invokeMethodRethrowing` was added to support this.
Lastly, there are a couple of small refactoring here and there to make the code more readable, and less error-prone.

### Steps for Testing
1. Log in to Artemis
2. Navigate to Course Administration
3. Create a new Java exercise
4. Check the created tests repo, which should include the new version of BehaviorTest class

### CODE REVIEW

**Comment by alllex** [comment]
Regarding using `clazz` in the feedback messages instead of a simple name. Using the class object itself calls `toString()` method, which provides more context on the actual class than a simple name. E.g. package name gets included, which helps to find the actual class faster. It could also indicate that something wrong is going on if the class in the error is not coming from the exercise but from the JDK. Moreover, it would nicely differentiate between classes and interfaces, again providing more context. On top of that, it would provide more explanatory message, if the class happens to be nested or anonymous, or lambda-based, etc. So I think, it is quite beneficial to have that, and it's also a bit shorter.

Sorry about the file version, I've trusted the experienced maintainers to look out for it.

The exercise where these changes have been used is this one:
https://artemis.ase.in.tum.de/#/course/41/programming-exercise/1057/view

**Comment by krusche** [comment]
> Regarding using `clazz` in the feedback messages instead of a simple name. Using the class object itself calls `toString()` method, which provides more context on the actual class than a simple name. E.g. package name gets included, which helps to find the actual class faster. It could also indicate that something wrong is going on if the class in the error is not coming from the exercise but from the JDK. Moreover, it would nicely differentiate between classes and interfaces, again providing more context. On top of that, it would provide more explanatory message, if the class happens to be nested or anonymous, or lambda-based, etc. So I think, it is quite beneficial to have that, and it's also a bit shorter.
> 
> Sorry about the file version, I've trusted the experienced maintainers to look out for it.
> 
> The exercise where these changes have been used is this one:
> https://artemis.ase.in.tum.de/#/course/41/programming-exercise/1057/view

Thanks for the explanation. We kept it intentionally "simple" using `clazz.getSimpleName()` because many students in our 1st and 2nd semester courses do not have much experience with all the details, so I am not sure if the additional context is really helping. Maybe some students do not understand this additional context? Also, please always keep in mind, that the feedback should not reveal too much of the actual solution.

File: src/main/resources/templates/java/test/testFiles/behavior/BehaviorTest.java [hunk 1]
```diff
@@ -40,18 +43,25 @@
...
-        String failMessage = "Could not instantiate the class '" + className + "' because";
+    /**
+     * Instantiate an object of a given class by its qualified name and the constructor arguments, if applicable.
+     * @param clazz: new instance of which is required
+     * @param constructorArgs: Parameter instances of the constructor of the class, that it has to get instantiated with. Do not include, if the constructor has no arguments.
+     * @return The instance of this class.
+     */
+    protected Object newInstance(Class<?> clazz, Object... constructorArgs) {
+        Class<?>[] constructorArgTypes = getParameterTypes(constructorArgs);
+        String failMessage = "Could not instantiate the " + clazz + " because";
```
Comment (krusche) [inline src/main/resources/templates/java/test/testFiles/behavior/BehaviorTest.java:64]: please revert this to print the class name

File: src/main/resources/templates/java/test/testFiles/behavior/BehaviorTest.java [hunk 2]
```diff
@@ -81,15 +91,53 @@ protected Object newInstance(String qualifiedClassName, Object... constructorArg
...
     }
 
+    /**
+     * Instantiate an object of a given class by its qualified name and the constructor arguments, if applicable.
+     * @param constructorArgs: Parameter instances of the constructor of the class, that it has to get instantiated with. Do not include, if the constructor has no arguments.
+     * @return The instance of this class.
+     */
+    protected Object newInstance(Constructor<?> constructor, Object... constructorArgs) {
+        String failMessage = "Could not instantiate the "
+                + constructor.getDeclaringClass() + " because";
```
Comment (krusche) [inline src/main/resources/templates/java/test/testFiles/behavior/BehaviorTest.java:101]: please use the class simple name here

File: src/main/resources/templates/java/test/testFiles/behavior/BehaviorTest.java [hunk 3]
```diff
@@ -81,15 +91,53 @@ protected Object newInstance(String qualifiedClassName, Object... constructorArg
...
     /**
      * Retrieve an attribute value of a given instance of a class by the attribute name.
      * @param object: The instance of the class that contains the attribute.
      * @param attributeName: The name of the attribute whose value needs to get retrieved.
      * @return The instance of the attribute with the wanted value.
      */
     protected Object valueForAttribute(Object object, String attributeName) {
-        String failMessage = "Could not retrieve the attribute '" + attributeName + "' from the class '"
-            + object.getClass().getSimpleName() + "' because";
+        requireNonNull(object, "receiver must not be null");
```
Comment (krusche) [inline src/main/resources/templates/java/test/testFiles/behavior/BehaviorTest.java:138]: I don't think that `requireNonNull`. How would the feedback in Artemis look like that students get to see in such a case? I think we should rather provide a good fail message.

File: src/main/resources/templates/java/test/testFiles/behavior/BehaviorTest.java [hunk 4]
```diff
@@ -113,6 +161,7 @@ protected Object valueForAttribute(Object object, String attributeName) {
      * @return The wanted method.
      */
     protected Method getMethod(Object object, String methodName, Class<?>... parameterTypes) {
+        requireNonNull(object, "receiver must not be null");
```
Comment (krusche) [inline src/main/resources/templates/java/test/testFiles/behavior/BehaviorTest.java:164]: I don't think that `requireNonNull`. How would the feedback in Artemis look like that students get to see in such a case? I think we should rather provide a good fail message.

File: src/main/resources/templates/java/test/testFiles/behavior/BehaviorTest.java [hunk 5]
```diff
@@ -125,10 +174,10 @@ protected Method getMethod(Object object, String methodName, Class<?>... paramet
      */
     protected Method getMethod(Class<?> declaringClass, String methodName, Class<?>... parameterTypes) {
         String failMessage = "Could not find the method '" + methodName + "' with the parameters: "
-            + getParameterTypesAsString(parameterTypes) + " from the class " + declaringClass.getSimpleName() + " because";
+            + getParameterTypesAsString(parameterTypes) + " from the " + declaringClass + " because";
```
Comment (krusche) [inline src/main/resources/templates/java/test/testFiles/behavior/BehaviorTest.java:177]: please revert to use the simple name of the declaring class

File: src/main/resources/templates/java/test/testFiles/behavior/BehaviorTest.java [hunk 6]
```diff
@@ -125,10 +174,10 @@ protected Method getMethod(Object object, String methodName, Class<?>... paramet
...
 
         if (parameterTypes == null || parameterTypes.length == 0) {
-            failMessage = "Could not find the method '" + methodName + "' from the class " + declaringClass.getSimpleName() + " because";
+            failMessage = "Could not find the method '" + methodName + "' from the " + declaringClass + " because";
```
Comment (krusche) [inline src/main/resources/templates/java/test/testFiles/behavior/BehaviorTest.java:180]: please revert to use the simple name of the declaring class

File: src/main/resources/templates/java/test/testFiles/behavior/BehaviorTest.java [hunk 7]
```diff
@@ -152,7 +201,9 @@ protected Method getMethod(Class<?> declaringClass, String methodName, Class<?>.
      * @return The return value of the method.
      */
     protected Object invokeMethod(Object object, Method method, Object... params) {
-        String failMessage = "Could not invoke the method '" + method.getName() + "' in the class '" + object.getClass().getSimpleName() + "' because";
+        // NOTE: object can be null, if method is static
+        String failMessage = "Could not invoke the method '" + method.getName()
+                + "' in the " + method.getDeclaringClass() + " because";
```
Comment (krusche) [inline src/main/resources/templates/java/test/testFiles/behavior/BehaviorTest.java:206]: please revert to use the simple name of the declaring class

File: src/main/resources/templates/java/test/testFiles/behavior/BehaviorTest.java [hunk 8]
```diff
@@ -180,26 +256,43 @@ protected Object invokeMethod(Object object, String methodName, Object... params
...
      */
-    private Class<?>[] getParameterTypes(Object... params) {
-        Class<?>[] parameterTypes;
-
-        if(params != null && params.length > 0) {
-            parameterTypes = new Class<?>[params.length];
+    protected Constructor<?> getConstructor(Class<?> declaringClass, Class<?>... parameterTypes) {
+        String failMessage = "Could not find the constructor with the parameters: "
+                + getParameterTypesAsString(parameterTypes)
+                + " from the " + declaringClass + " because";
```
Comment (krusche) [inline src/main/resources/templates/java/test/testFiles/behavior/BehaviorTest.java:267]: please use simpleName here

File: src/main/resources/templates/java/test/testFiles/behavior/BehaviorTest.java [hunk 9]
```diff
@@ -180,26 +256,43 @@ protected Object invokeMethod(Object object, String methodName, Object... params
...
 
-            for(Object param : params) {
-                Class<?> paramType = param.getClass();
-                parameterTypes[Arrays.asList(params).indexOf(param)] = paramType;
-            }
+        if (parameterTypes == null || parameterTypes.length == 0) {
+            failMessage = "Could not find the constructor from the " + declaringClass + " because";
```
Comment (krusche) [inline src/main/resources/templates/java/test/testFiles/behavior/BehaviorTest.java:270]: please use simple name here

File: src/main/resources/templates/java/test/testFiles/behavior/BehaviorTest.java [hunk 10]
```diff
@@ -180,26 +256,43 @@ protected Object invokeMethod(Object object, String methodName, Object... params
...
+    }
+
+    /**
+     * Retrieves the parameters types of a given collection of parameter instances.
+     * @param params: The instances of the parameters.
+     * @return The parameter types of the instances as an array.
+     */
+    private Class<?>[] getParameterTypes(Object... params) {
+        return Arrays.stream(params)
+                .map(it -> requireNonNull(it, "parameters must not be null"))
```
Comment (krusche) [inline src/main/resources/templates/java/test/testFiles/behavior/BehaviorTest.java:293]: see discussion about requireNonNull above

**Comment by krusche** [comment]
@MaisiKoleni can you take over this PR? I think there are some valid improvements, but some of the changes need to be reverted.

**Comment by MaisiKoleni** [comment]
Closing this pull request as it is being continued in #1442.

**Pull request closed without merging**